tamr_client/datasets/record.py

```diff
@@ -0,0 +1,124 @@
+"""
+See https://docs.tamr.com/reference/record
+"The recommended approach for interacting with records is to use the upsert and delete functions for all use cases they
+can handle. For more advanced use cases, the underlying _update function can be used directly."
+"""
+import json
+from typing import Dict, Iterable, Union
+
+import tamr_client as tc
+from tamr_client.types import JsonDict
+
+
+class PrimaryKeyNotFound(Exception):
+    """Raised when referencing a primary key by name that does not exist."""
+
+    pass
+
+
+def _update(
+    session: tc.Session, dataset: tc.Dataset, updates: Iterable[Dict]
+) -> JsonDict:
+    """Send a batch of record creations/updates/deletions to this dataset.
+    You probably want to use :func:`~tamr_client.dataset.upsert_records`
+    or :func:`~tamr_client.dataset.delete_records` instead.
+
+    Args:
+        dataset: Dataset containing records to be updated
+        updates: Each update should be formatted as specified in the `Public Docs for Dataset updates <https://docs.tamr.com/reference#modify-a-datasets-records>`_.
+
+    Returns:
+        JSON response body from server
+
+    Raises:
+        requests.HTTPError: If an HTTP error is encountered
+    """
+    stringified_updates = (json.dumps(update) for update in updates)
+    r = session.post(
+        str(dataset.url) + ":updateRecords",
+        headers={"Content-Encoding": "utf-8"},
+        data=stringified_updates,
+    )
+    return tc.response.successful(r).json()
+
+
+def upsert(
+    session: tc.Session,
+    dataset: tc.Dataset,
+    records: Iterable[Dict],
+    *,
+    primary_key_name: str,
+) -> JsonDict:
+    """Create or update the specified records.
+
+    Args:
+        dataset: Dataset to receive record updates
+        records: The records to update, as dictionaries
+        primary_key_name: The primary key for these records, which must be a key in each record dictionary
+
+    Returns:
+        JSON response body from server
+
+    Raises:
+        requests.HTTPError: If an HTTP error is encountered
+        PrimaryKeyNotFound: If primary_key_name does not match dataset primary key
+        PrimaryKeyNotFound: If primary_key_name not in a record dictionary
+    """
+    if primary_key_name not in dataset.key_attribute_names:
```

Comment:
the API supports datasets with composite keys; tamr-client should support them as well

Reply:
We did a quick spike to investigate composite keys. Apparently, the API accepts composite keys, but the backend can't handle them, so we shouldn't give users the ability to until all parts of the API use composite keys and the backend can actually do meaningful things with composite keys.

Reply:
@skalish might be worth at least documenting in the module-level docstring that we don't handle composite keys

Reply:
tamr-client interface should only change in backward-compatible ways to introduce composite key support; right now, primary_key_name is type str - options in the future are to introduce another parameter, e.g. primary_key_names of type list of str, or to make primary_key_name the union of str and list of str

Reply:
Conversation moved to #349